concent_api/utils/message_handler.py

```diff
@@ -0,0 +1,72 @@
+from functools                      import wraps
+
+from django.db                      import transaction
+from django.http                    import JsonResponse
+from django.http                    import HttpResponse
+from django.http                    import HttpResponseNotAllowed
+from django.conf                    import settings
+
+from golem_messages                 import dump
+from golem_messages.message         import Message
+
+from core.exceptions                import Http400
+
+from utils                          import logging
+
+
+def message_handler(view):
+    @wraps(view)
+    def wrapper(request, message, *args, **kwargs):
+        try:
+            sid = transaction.savepoint()
+            response_from_view = view(request, message, *args, **kwargs)
+            transaction.savepoint_commit(sid)
+        except Http400 as exception:
+            logging.log_400_error(
+                view.__name__,
+                message,
+                message.client_public_key,
+            )
+            transaction.savepoint_rollback(sid)
+            return JsonResponse({'error': str(exception)}, status = 400)
+        if isinstance(response_from_view, Message):
+            assert response_from_view.sig is None
+            logging.log_message_returned(
+                response_from_view,
+                message.client_public_key,
+            )
+            serialized_message = dump(
+                response_from_view,
+                settings.CONCENT_PRIVATE_KEY,
+                message.client_public_key,
+            )
+
+            return HttpResponse(serialized_message, content_type = 'application/octet-stream')
+        elif isinstance(response_from_view, dict):
+            return JsonResponse(response_from_view, safe = False)
+        elif isinstance(response_from_view, HttpResponseNotAllowed):
+            logging.log_message_not_allowed(
+                view.__name__,
+                request.method,
+                message.client_public_key,
+            )
+            return response_from_view
+        elif isinstance(response_from_view, HttpResponse):
+            logging.log_message_accepted(
+                message,
+                message.client_public_key,
+            )
+            return response_from_view
+        elif response_from_view is None:
+            logging.log_empty_queue(
+                view.__name__,
+                message.client_public_key,
+            )
+            return HttpResponse("", status = 204)
+        elif isinstance(response_from_view, bytes):
+            return HttpResponse(response_from_view)
+
+        assert False, "Invalid response type"
+        raise Exception("Invalid response type")
```

Comment:
This code is unreachable. I would delete `assert` above ;)

Reply:
It's been added here only because pylint complained about missing return.

If anything, we could delete the `raise` but I think it's better than disabling the warning.

Reply:
Because this part is never meant to run in production. Code that would reach this `raise` is so broken that it should never make it to production. Doing anything here is simply not necessary unless we want to code defensively and add additional logic to handle "what if" situations resulting from bugs.

This `assert` is like a white line on the street. It informs us that this is a boundary that should not be crossed and makes it apparent that we stepped past it. But that's it. It's not meant to hold our backs if we actually cross it in a real situation.

To be honest this case is borderline because the "handling" is simply a single `raise` so it's not that different from `assert False`. As long as it's not just a silent `return`, doing it instead of `assert` is not harmful, just unnecessary in my opinion. It would be different if we needed more code to handle this. Anything more than an `assert` would be a waste of time.

For me `assert` is an important tool to avoid coding defensively - i.e. for every nonsensical case I can think of - while still being sure that if I'm wrong or if the code around changes this will not go unnoticed. I can just write an `assert` that makes it clear that I'm sure that a situation is impossible (or a result of a bug elsewhere) and move on to more important parts of the code. In other words, please don't think about it as assert vs no assert. More often it's assert vs lots of unnecessary code to handle something that cannot happen.